### DISCUSSION TITLE
Request Error: unable to get local issuer certificate

### DISCUSSION THREAD

**Discussion by lotek4u** [community/community#36894] [Enterprise] [answered]
I've seen many people with this issue, but no full solutions posted. I'm working at an enterprise shop and we have proxys that perform SSL inspection. It's a fact of life. There is nothing I can do about... but I can put in bypass rules. I had our proxy team disable SSL inspection for https://vscode.dev/codeExchangeProxyEndpoints/github/login/oauth/access_token
which allowed me to activate the copilot extension, but I still can't use copilot.

[2022-10-21 08:20:04.801] [renderer1] [error] [Extension Host] [ERROR] [default] [2022-10-21T13:20:04.797Z] Request Error: unable to get local issuer certificate
[2022-10-21 08:20:04.801] [renderer1] [error] [Extension Host] [ERROR] [ghostText] [2022-10-21T13:20:04.797Z] Error on ghost text request FetchError: unable to get local issuer certificate

I have github auth extension enabled, my github and msft accounts are both active in Code, but copilot is failing with a cert error. I've uninstalled Code, removed the .vscode and %appdata% folders from my windows machine, reinstalled Code and just used those 2 extensions... still not working.

What exactly is the problem now? Is there another URL I need to disable SSL inspection with? It was difficult to even find that URL, it would be very helpful if someone documented what these errors mean and how to resolve them. Just saying "there is probably a VPN or something breaking it" does not help at all. What is the solution?  

Thanks,
Nick

**Comment by iamcjwilliams** [comment]
@lotekjunky Thank you for your feedback, our copilot team is investigating this

**Reply by lotek4u** [reply]
Is there an update on this? I can't use it at work.

**Comment by lotek4u** [comment]
This morning, I tried copilot out again. I'm not longer getting the local issuer certificate error and copilot is giving me AI suggestions as designed. As far as I know, nothing has changed on my side, but it's working now.

**Comment by lotek4u** [comment]
Hey, it's broken again. Worked fine last week, nothing changed on my side... copilot activation is disabled again.

**Reply by iamcjwilliams** [reply]
@lotekjunky  can you provide a screenshot of the error you're receiving by performing the following: (View -> Output; select GitHub Copilot from the pulldown)

**Reply by lotek4u** [reply]
Here you go @iamcjwilliams :
https://imgur.com/a/Mh00re1

[ERROR] [default] [2022-11-01T13:26:29.021Z] GitHub Copilot could not connect to server. Extension activation failed: "unable to get local issuer certificate"

**Reply by iamcjwilliams** [reply]
The error message unable to get local issuer certificate means a proxy that uses self-signed certificates is being used. Unfortunately, self-signed certificate is something we currently don't have support for but will soon. I recommend that you keep an eye on our [official GitHub public product roadmap](https://github.com/github/roadmap), our [Changelog](https://github.blog/changelog/), and our [GitHub Blog](http://github.blog/) for future updates.

**Reply by lotek4u** [reply]
thank you @iamcjwilliams .  I will keep an eye on that.  However, I had this working at work last week... and I have the ability to put in SSL bypass for certain URLs in the corporate proxy.  I just can't find what needs to be bypassed.  I was able to get github authentication working, but the copilot activation is what's failing "obviously."

**Comment by chuklee** [comment]
Hello any news on this subject, i have the same error :/

**Reply by lotek4u** [reply]
I never got anywhere with it and cancelled copilot. They say if you want to use it with a proxy, get an enterprise license.

**Reply by lansalot** [reply]
Well, I should have found this thread before I signed up I guess! :(

**Reply by 0xant** [reply]
true,

**Reply by 0xant** [reply]
cancelling the subscription,

**Reply by FarisAhmed** [reply]
@lotekjunky That is not true, I managed to make it work with Visual Studio 2022 Enterprise behind Zscaler on my corporate notebook

**Reply by Shurid** [reply]
Where do I find these proxy settings in Visual Studio 2022?

**Reply by FarisAhmed** [reply]
You don't need to change any proxy settings. Just follow the steps in my post. It worked for me and for many others too

**Reply by Shurid** [reply]
Thanks it worked.

**Comment by nklsla** [comment, marked as answer]
Think I got it working. I'm behind company proxy and these are my settings I believe are relevant:
```
    "github.copilot.advanced": {
        "authProvider": "github"
    },
    "http.proxySupport": "on",
    "http.proxy": "http://<company-url>:<port>",
    "http.proxyStrictSSL": false,
```
When I changed `http.proxyStrictSSL` to `false` the `local issure certificate` problem was resolved.

Also worth mentioning is that I've marked Copilot as `Trusted` in `Manage Trusted Extensions` in my github profile in Vscode (found at the bottom left where you activate settings sync)
@hem4n7 @lansalot @lotekjunky @chuklee

**Reply by Depechie** [reply]
This worked for me too!

**Reply by marco-kasl** [reply]
Thanks @nklsla!
It worked perfectly for me!

**Reply by Shurid** [reply]
With FarisAhmed solution I was able to solve certification issue but still having corporate proxy issue. If I disconnect VPN everything works but with VPN it gives error: Error while processing the event: The proxy tunnel request to proxy 'http://xxxx/' failed with status code '407'." Where do I find the setting file to update as you have stated above? I am not able to locate the settings that needs the change to resolve.

**Reply by nklsla** [reply]
I had that issue for the Chat-function, but if I remember correctly it got solved when I updated vs code to the latest version.

**Reply by lotek4u** [reply]
I decided to resubscribe and tested this out and it actually worked! Thanks @marco-kasl !  I'm having issues with the chat still doing the same thing, but I can live with that for now I guess...

**Reply by ataberkozek** [reply]
worked perfectly thanks

**Reply by fitlemon** [reply]
I deleted suspucious keys, and keep remained only two of them in copilot settings and it works:
```
"github.copilot.advanced": {
        "authProvider": "github",
        "debug.useNodeFetcher": true,
    },

**Reply by ponchofiesta** [reply]
I don't want to disable certificate checks using `"http.proxyStrictSSL": false,` so there seems to be no secure solution.

**Comment by ManuelYepiz** [comment]
Hi, I appreciate your help @nklsla, I was unable to apply your suggestion, as I didn't get it to be honest v: haha, but I got this suggestion from a coworker that fixed it for me:

"Please, try this:
set NODE_EXTRA_CA_CERTS system environment variable to refer to your git certificate bundle file, usually ‘c:\Program Files\Git\mingw64\etc\ssl\certs\ca-bundle.trust.crt’
(this demands git installed on your computer)"

For anyone not familiar with this menu:
![image](https://github.com/community/community/assets/149177760/d8aec9ac-b437-4250-9e6c-3d16e65a263d)
![image](https://github.com/community/community/assets/149177760/a591ff6d-c12c-4031-8e79-3c5fb7cdea96)

Then just fill the boxes with NODE_EXTRA_CA_CERTS & the Path.

Hopefully this is useful for others, I can get suggestions now by adding a comment prior to the line where I want the code:
![image](https://github.com/community/community/assets/149177760/6e321ed5-8c5c-4e3f-9759-391afaab3fd7)
 

Regards.

**Reply by angelhortelano** [reply]
This worked perfectly for me, thank you very much

**Reply by lotek4u** [reply]
This one didn't work for me. I could never make VS Code care about the fact that I added a new ENV variable.   It's up to you if you want to provide more info, but it might help others later... it would be very useful to see your Github Copilot Output box and/or Copilot config section of settings.json

![image](https://github.com/community/community/assets/29106552/50155938-f2b8-46be-ad78-d91b6418d439)

**Comment by JSantas** [comment]
If you are using copilot with zscaler, this could help with **unable to get local issuer cert**
<img width="840" alt="image" src="https://github.com/community/community/assets/68399093/e44ff3c7-f318-433a-9cda-12729d14fa1f">

**Reply by meChiel** [reply]
Hey, 
I am trying to use using copilot with zscaler. And I am having the "unable to get local issuer cert" issue.
So I am interested in this solution to make it work. But I am unable to find the settings dialogue box you show.
Where, in what program, how do you find these settings?

**Reply by mawax** [reply]
Looks like IntelliJ. But similar settings can be found in Visual Studio Code

**Comment by FarisAhmed** [comment]
I managed to make my 30 days free trial CoPilot work in Visual Studio 2022 on Windows behind Zscaler as follows:

* Browse with chrome to https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex then export the Zscaler Root CA as shown below to file C:\Data\CoPilot\CA\Zscaler Root CA.pem
![image](https://github.com/community/community/assets/8459423/e9b42d3a-ba04-432f-a8c2-c016d1d32a0a)

* Define new user environment variable NODE_EXTRA_CA_CERTS and set its value to the Zscaler certificate file
* Restart Visual Studio

**Comment by AI-FireSnake** [comment]
Hi,

anyone got the following error? On Visual Studio 2022 release 17.10.2? Thanks

[CopilotClientManager Error] [default] Error sending telemetry FetchError: unable to get local issuer certificate
    at fetch (C:\snapshot\copilot-client\node_modules\@adobe\helix-fetch\src\fetch\index.js:99:11)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at cachingFetch (C:\snapshot\copilot-client\node_modules\@adobe\helix-fetch\src\fetch\index.js:288:16)
    at cX.fetch (C:\snapshot\copilot-client\lib\src\network\helix.ts:78:22) {
  type: 'system',
  _name: 'FetchError',
  code: 'UNABLE_TO_GET_ISSUER_CERT_LOCALLY',
  errno: undefined,
  erroredSysCall: undefined
}

**Comment by allanmnolasco** [comment]
Hello everyone, 
I have different settings and copilot chat it's working well, are here if you want to try it.

by the way this configuration is using Zscaler.

 `"http.proxyStrictSSL": false,
    "http.proxyAuthorization": null,
    "http.proxySupport": "override",
    "http.systemCertificates": true,
    "github.copilot.advanced": {
        "debug.useNodeFetcher": true
    },`

**Reply by yardobr** [reply]
bro, you just saved me! nothing else helped me and I lost the hope 😆 
thank you a lot!

**Reply by haerriz** [reply]
Where do I have to update this to make the github copilot chat work?

**Reply by yardobr** [reply]
these are your VSCode settings

**Comment by Shubhk007** [comment]
I found another way of doing it. You have to check if it works for you- 

1. Install win-ca from VS marketplace
2. Go to win-ca settings
3. Change win-ca method from **inject** to **append**.

Restart VS code. Worked for me :)

**Reply by ponchofiesta** [reply]
Did not work for me.
